--- ./modules/core/src/va_intel.cpp.orig	2018-12-02 02:08:43.105140530 +0000
+++ ./modules/core/src/va_intel.cpp	2018-12-02 02:44:51.917011656 +0000
@@ -340,8 +340,8 @@ static void copy_convert_yv12_to_bgr(con
             1.5959997177f
         };
 
-    CV_CheckEQ(image.format.fourcc, VA_FOURCC_YV12, "Unexpected image format");
-    CV_CheckEQ(image.num_planes, 3, "");
+    //CV_CheckEQ(image.format.fourcc, VA_FOURCC_YV12, "Unexpected image format");
+    //CV_CheckEQ(image.num_planes, 3, "");
 
     const size_t srcOffsetY = image.offsets[0];
     const size_t srcOffsetV = image.offsets[1];
@@ -417,8 +417,8 @@ static void copy_convert_bgr_to_yv12(con
             -0.2909994125f, 0.438999176f, -0.3679990768f, -0.0709991455f
         };
 
-    CV_CheckEQ(image.format.fourcc, VA_FOURCC_YV12, "Unexpected image format");
-    CV_CheckEQ(image.num_planes, 3, "");
+    //CV_CheckEQ(image.format.fourcc, VA_FOURCC_YV12, "Unexpected image format");
+    //CV_CheckEQ(image.num_planes, 3, "");
 
     const size_t dstOffsetY = image.offsets[0];
     const size_t dstOffsetV = image.offsets[1];
