--- OpenCV-2.0.0/data/Makefile.am.orig	2009-10-01 01:20:58.000000000 +0100
+++ OpenCV-2.0.0/data/Makefile.am	2009-12-04 13:05:11.000000000 +0100
@@ -25,4 +25,6 @@
   haarcascades/haarcascade_mcs_mouth.xml \
   haarcascades/haarcascade_mcs_nose.xml \
   haarcascades/haarcascade_mcs_righteye.xml \
-  haarcascades/haarcascade_mcs_upperbody.xml
+  haarcascades/haarcascade_mcs_upperbody.xml \
+  \
+  lbpcascades/lbpcascade_frontalface.xml
