diff -up opencv-1.1.0/configure.in.conflicts opencv-1.1.0/configure.in
--- opencv-1.1.0/configure.in.conflicts	2009-04-24 12:01:24.000000000 +0200
+++ opencv-1.1.0/configure.in	2009-04-24 12:01:29.000000000 +0200
@@ -37,7 +37,7 @@ AC_PREREQ(2.60)
 AC_INIT([opencv],[opencv_version],[opencvlibrary-devel@lists.sourceforge.net])
 AC_CONFIG_SRCDIR([opencv.pc.in])
 AC_CONFIG_AUX_DIR([autotools])
-AC_CONFIG_MACRO_DIR([autotools/aclocal])
+
 
 # put CVS revision of this configure.in into the resulting configure script
 AC_REVISION($Revision: 1.60 $)
