diff -urNp old/CMakeLists.txt new/CMakeLists.txt
--- old/CMakeLists.txt	2016-12-23 13:54:44.000000000 +0100
+++ new/CMakeLists.txt	2017-02-27 08:56:04.504933013 +0100
@@ -402,12 +402,7 @@ else()
     ocv_update(OPENCV_OTHER_INSTALL_PATH   share/OpenCV)
 
     if(NOT DEFINED OPENCV_CONFIG_INSTALL_PATH)
-      math(EXPR SIZEOF_VOID_P_BITS "8 * ${CMAKE_SIZEOF_VOID_P}")
-      if(LIB_SUFFIX AND NOT SIZEOF_VOID_P_BITS EQUAL LIB_SUFFIX)
-        ocv_update(OPENCV_CONFIG_INSTALL_PATH lib${LIB_SUFFIX}/cmake/opencv)
-      else()
-        ocv_update(OPENCV_CONFIG_INSTALL_PATH share/OpenCV)
-      endif()
+      ocv_update(OPENCV_CONFIG_INSTALL_PATH lib${LIB_SUFFIX}/OpenCV)
     endif()
   endif()
   ocv_update(OPENCV_INCLUDE_INSTALL_PATH "include")
