diff -urp opencv-0.9.7.orig/interfaces/swig/python/Makefile.am opencv-0.9.7/interfaces/swig/python/Makefile.am
--- opencv-0.9.7.orig/interfaces/swig/python/Makefile.am	2006-03-06 22:58:41.000000000 -0500
+++ opencv-0.9.7/interfaces/swig/python/Makefile.am	2006-03-07 09:42:55.000000000 -0500
@@ -18,7 +18,7 @@ AM_CPPFLAGS = \
 
 if BUILD_PYTHON_WRAPPERS
 
-  pkgpython_LTLIBRARIES = _cv.la _highgui.la
+  pkgpyexec_LTLIBRARIES = _cv.la _highgui.la
 
   pkgpython_PYTHON      = \
     __init__.py    \
