--- OpenCV-2.0.0/apps/Makefile.am.orig	2009-10-01 01:20:58.000000000 +0100
+++ OpenCV-2.0.0/apps/Makefile.am	2009-12-04 13:36:09.000000000 +0100
@@ -4,11 +4,16 @@ AM_LDFLAGS  = @OPENMP_CXXFLAGS@
 
 EXTRA_DIST = \
   CMakeLists.txt \
-  haartraining/CMakeLists.txt 
+  haartraining/CMakeLists.txt \
+  transcascade/CMakeLists.txt
 
 noinst_LTLIBRARIES = libcvhaartraining.la
 
-bin_PROGRAMS = opencv-haartraining opencv-createsamples opencv-performance
+bin_PROGRAMS = \
+  opencv-haartraining \
+  opencv-createsamples \
+  opencv-performance \
+  opencv-traincascade
 
 ALL_CV_LIBS = \
   $(top_builddir)/src/libhighgui.la       \
@@ -27,7 +32,12 @@ noinst_HEADERS = \
   haartraining/cvclassifier.h    \
   haartraining/cvhaartraining.h  \
   haartraining/_cvhaartraining.h \
-  haartraining/_cvcommon.h
+  haartraining/_cvcommon.h \
+  traincascade/boost.h \
+  traincascade/cascadeclassifier.h \
+  traincascade/features.h \
+  traincascade/imagestorage.h \
+  traincascade/lbpfeatures.h
 
 opencv_haartraining_SOURCES = haartraining/haartraining.cpp
 opencv_haartraining_LDADD = libcvhaartraining.la $(ALL_CV_LIBS)
@@ -38,3 +48,14 @@ opencv_createsamples_LDADD = libcvhaartr
 opencv_performance_SOURCES = haartraining/performance.cpp
 opencv_performance_LDADD = libcvhaartraining.la $(ALL_CV_LIBS)
 
+opencv_traincascade_SOURCES = \
+  traincascade/traincascade.cpp \
+  traincascade/cascadeclassifier.cpp \
+  traincascade/boost.cpp \ 
+  traincascade/features.cpp \
+  traincascade/haarfeatures.cpp \
+  traincascade/lbpfeatures.cpp \
+  traincascade/imagestorage.cpp 
+opencv_traincascade_LDADD = \
+  $(ALL_CV_LIBS) \
+  $(top_builddir)/src/libml.la
