diff -urp opencv-0.9.7.orig/cxcore/include/cxtypes.h opencv-0.9.7/cxcore/include/cxtypes.h
--- opencv-0.9.7.orig/cxcore/include/cxtypes.h	2006-03-06 22:58:41.000000000 -0500
+++ opencv-0.9.7/cxcore/include/cxtypes.h	2006-03-06 22:59:56.000000000 -0500
@@ -48,7 +48,7 @@
   #include <string.h>
   #include <float.h>
 
-  #if defined WIN64 && defined EM64T && defined _MSC_VER
+  #if defined WIN64 && defined EM64T && defined _MSC_VER || defined __SSE2__
     #include <emmintrin.h>
   #endif
 
