diff -up opencv-1.1.0/configure.in.nooptim opencv-1.1.0/configure.in
--- opencv-1.1.0/configure.in.nooptim	2008-10-15 17:47:43.000000000 +0200
+++ opencv-1.1.0/configure.in	2009-04-23 18:54:30.000000000 +0200
@@ -274,41 +274,41 @@ else
     AC_MSG_NOTICE( TARGET=$target )
     case $target in
     i*86-apple-*)
-        if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+        if test x"no" = "xyes"; then
             # apples g++ fails with '-march=i686' and there are no apple machines older than prescott/core anyway
             DEFAULT_CXXFLAGS="-g -march=prescott -ffast-math -fomit-frame-pointer $DEFAULT_CXXFLAGS"
         fi
         ;;
     i686-*-*)
-        if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+        if test x"no" = "xyes"; then
           # default to i686/pentiumpro -- people can override this
           DEFAULT_CXXFLAGS="-g -march=i686 -ffast-math -fomit-frame-pointer $DEFAULT_CXXFLAGS"
         fi
         ;;
     amd64-*-* | x86_64-*-*)
-        if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+        if test x"no" = "xyes"; then
           # default to i686/pentiumpro -- people can override this
           DEFAULT_CXXFLAGS="-g $DEFAULT_CXXFLAGS"
         fi
         ;;    
     powerpc-*-*)
-        if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+        if test x"no" = "xyes"; then
             # default to G3 processors -- people can override this
             DEFAULT_CXXFLAGS="-g -mcpu=G3 -mtune=G5 -fomit-frame-pointer $DEFAULT_CXXFLAGS"
         fi
         ;;
     *-*-*)
-        if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+        if test x"no" = "xyes"; then
             DEFAULT_CXXFLAGS="-fomit-frame-pointer $DEFAULT_CXXFLAGS"
         fi
         ;;
     esac
 
     CPPFLAGS="-DNDEBUG $CPPFLAGS"
-    DEFAULT_CXXFLAGS="-O3 $DEFAULT_CXXFLAGS"
+#    DEFAULT_CXXFLAGS="-O3 $DEFAULT_CXXFLAGS"
 fi
 
-if test x"$ac_cv_c_compiler_gnu" = "xyes"; then
+if test x"no" = "xyes"; then
     PY_CXXFLAGS="-Wall -pipe $DEFAULT_CXXFLAGS"
     OCT_CXXFLAGS="-fno-strict-aliasing -Wall -Wno-uninitialized -pipe $DEFAULT_CXXFLAGS"
     DEF_CXXFLAGS="-Wall -fno-rtti -pipe $DEFAULT_CXXFLAGS"
