diff -up OpenCV-2.4.4/modules/python/test/ticket_6.py.pillow OpenCV-2.4.4/modules/python/test/ticket_6.py
--- OpenCV-2.4.4/modules/python/test/ticket_6.py.pillow	2013-01-28 00:00:48.000000000 +0100
+++ OpenCV-2.4.4/modules/python/test/ticket_6.py	2013-02-18 19:58:14.288877399 +0100
@@ -2,7 +2,7 @@
 
 import urllib
 import cv2.cv as cv
-import Image
+from PIL import Image
 import unittest
 
 class TestLoadImage(unittest.TestCase):
