diff -up OpenCV-2.4.3/modules/highgui/src/cap_gstreamer.cpp.codecs OpenCV-2.4.3/modules/highgui/src/cap_gstreamer.cpp
--- OpenCV-2.4.3/modules/highgui/src/cap_gstreamer.cpp.codecs	2012-09-06 00:55:12.000000000 +0200
+++ OpenCV-2.4.3/modules/highgui/src/cap_gstreamer.cpp	2013-01-06 00:14:14.628006270 +0100
@@ -442,12 +442,19 @@ protected:
 
 void CvVideoWriter_GStreamer::init()
 {
-    encs[CV_FOURCC('H','F','Y','U')]=(char*)"ffenc_huffyuv";
     encs[CV_FOURCC('D','R','A','C')]=(char*)"diracenc";
-    encs[CV_FOURCC('X','V','I','D')]=(char*)"xvidenc";
-    encs[CV_FOURCC('X','2','6','4')]=(char*)"x264enc";
+    encs[CV_FOURCC('H','F','Y','U')]=(char*)"ffenc_huffyuv";
+    encs[CV_FOURCC('J','P','E','G')]=(char*)"jpegenc";
+    encs[CV_FOURCC('M','J','P','G')]=(char*)"jpegenc";
     encs[CV_FOURCC('M','P','1','V')]=(char*)"mpeg2enc";
-    //encs[CV_FOURCC('M','P','2','V')]=(char*)"mpeg2enc";
+    encs[CV_FOURCC('M','P','2','V')]=(char*)"mpeg2enc";
+    encs[CV_FOURCC('T','H','E','O')]=(char*)"theoraenc";
+    encs[CV_FOURCC('V','P','8','0')]=(char*)"vp8enc";
+    encs[CV_FOURCC('H','2','6','4')]=(char*)"x264enc";
+    encs[CV_FOURCC('X','2','6','4')]=(char*)"x264enc";
+    encs[CV_FOURCC('X','V','I','D')]=(char*)"xvidenc";
+    encs[CV_FOURCC('F','F','Y','U')]=(char*)"y4menc";
+    //encs[CV_FOURCC('H','F','Y','U')]=(char*)"y4menc";
     pipeline=0;
     buffer=0;
 }
