Index: OpenCV-2.4.3/modules/contrib/doc/facerec/src/crop_face.py
===================================================================
--- OpenCV-2.4.3.orig/modules/contrib/doc/facerec/src/crop_face.py
+++ OpenCV-2.4.3/modules/contrib/doc/facerec/src/crop_face.py
@@ -31,7 +31,8 @@
 # ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
-import sys, math, Image
+import sys, math
+from PIL import Image
 
 def Distance(p1,p2):
   dx = p2[0] - p1[0]
Index: OpenCV-2.4.3/modules/python/test/ticket_6.py
===================================================================
--- OpenCV-2.4.3.orig/modules/python/test/ticket_6.py
+++ OpenCV-2.4.3/modules/python/test/ticket_6.py
@@ -1,6 +1,6 @@
 import urllib
 import cv2.cv as cv
-import Image
+from PIL import Image
 import unittest
 
 class TestLoadImage(unittest.TestCase):
